buildpacks/nodejs-engine/src/bin/web_env.rs

```diff
@@ -0,0 +1,79 @@
+use libcnb::data::exec_d::ExecDProgramOutputKey;
+use libcnb::data::exec_d_program_output_key;
+use libcnb::exec_d::write_exec_d_program_output;
+use std::cmp;
+use std::collections::HashMap;
+use std::env;
+use std::fs;
+
+const MAX_AVAILABLE_MEMORY_MB: usize = 14336;
+const DEFAULT_AVAILABLE_MEMORY_MB: usize = 512;
+const DEFAULT_WEB_MEMORY_MB: usize = 512;
+
+pub fn main() {
+    write_exec_d_program_output(web_env());
+}
+
+fn web_env() -> HashMap<ExecDProgramOutputKey, String> {
+    let available_mem = detect_available_memory();
+    let web_mem = detect_web_memory();
+    let web_concurrency = calculate_web_concurrency(web_mem, available_mem);
+
+    HashMap::from([
+        (
+            exec_d_program_output_key!("WEB_CONCURRENCY"),
+            web_concurrency.to_string(),
+        ),
+        (
+            exec_d_program_output_key!("WEB_MEMORY"),
+            web_mem.to_string(),
+        ),
+    ])
+}
+
+fn calculate_web_concurrency(web_mem: usize, available_mem: usize) -> usize {
+    let concurrency = available_mem / web_mem;
+    cmp::max(1, concurrency)
+}
+
+fn detect_available_memory() -> usize {
+    fs::read_to_string("/sys/fs/cgroup/memory.max")
+        .or_else(|_| fs::read_to_string("/sys/fs/cgroup/memory/memory.limit_in_bytes"))
+        .ok()
+        .map(|m| m.trim().to_string())
+        .and_then(|m| m.parse::<usize>().ok())
+        .map(|m| m / 1_048_576)
+        .map_or(DEFAULT_AVAILABLE_MEMORY_MB, |m| {
+            cmp::min(m, MAX_AVAILABLE_MEMORY_MB)
+        })
+}
```

Comment:
FWIW, another refactoring suggestion:

```suggestion
fn detect_available_memory() -> usize {
    [
        "/sys/fs/cgroup/memory.max",
        "/sys/fs/cgroup/memory/memory.limit_in_bytes",
    ]
    .iter()
    .find_map(|path| fs::read_to_string(path).ok())
    .and_then(|contents| contents.trim().parse().ok())
    .map_or(DEFAULT_AVAILABLE_MEMORY_MB, |limit_in_bytes: usize| {
        cmp::min(MAX_AVAILABLE_MEMORY_MB, limit_in_bytes / 1_048_576)
    })
}
```

Reply:
Fixed in [d701275](https://github.com/heroku/buildpacks-nodejs/pull/184/commits/d701275134b0f4aa3a1ef91cd1a5b01adb6bb727)